File: src/clang/magma/compile/rule/InfixRule.c
package magma.compile.rule;class package magma.compile.rule;{package magma.compile.rule;

import magma.compile.CompileError;class import magma.compile.CompileError;{

import magma.compile.CompileError;
import magma.compile.Node;class import magma.compile.Node;{
import magma.compile.Node;
import magma.result.Result;class import magma.result.Result;{
import magma.result.Result;

public record InfixRule(Rule left, String infix, Rule right) implements  Rule {
    @Override
    public Result<Node, CompileError> parse(String input) {
        return left.parse(input).and(() -> right.parse(input)).mapValue(tuple -> tuple.left().merge(tuple.right()));
    }

    @Override
    public Result<String, CompileError> generate(Node node) {
        return left.generate(node).and(() -> right.generate(node)).mapValue(tuple -> tuple.left() + infix + tuple.right());
    }
}
class public record InfixRule(Rule left, String infix, Rule right) implements  Rule {
    @Override
    public Result<Node, CompileError> parse(String input) {
        return left.parse(input).and(() -> right.parse(input)).mapValue(tuple -> tuple.left().merge(tuple.right()));
    }

    @Override
    public Result<String, CompileError> generate(Node node) {
        return left.generate(node).and(() -> right.generate(node)).mapValue(tuple -> tuple.left() + infix + tuple.right());
    }
}{

public record InfixRule(Rule left, String infix, Rule right) implements  Rule {
    @Override
    public Result<Node, CompileError> parse(String input) {
        return left.parse(input).and(() -> right.parse(input)).mapValue(tuple -> tuple.left().merge(tuple.right()));
    }

    @Override
    public Result<String, CompileError> generate(Node node) {
        return left.generate(node).and(() -> right.generate(node)).mapValue(tuple -> tuple.left() + infix + tuple.right());
    }
}

File: src/java/magma/app/compile/InfixRule.java
package magma.app.compile;

public final class InfixRule<Node, Error, NodeResult extends MergeNodeResult<Node, NodeResult>, StringResult extends AppendableStringResult<StringResult>> implements
        Rule<Node, NodeResult, StringResult> {
    private final Rule<Node, NodeResult, StringResult> leftRule;
    private final String infix;
    private final Rule<Node, NodeResult, StringResult> rightRule;
    private final ResultFactory<Node, Error, NodeResult, StringResult> factory;

    public InfixRule(Rule<Node, NodeResult, StringResult> leftRule, String infix, Rule<Node, NodeResult, StringResult> rightRule, ResultFactory<Node, Error, NodeResult, StringResult> factory) {
        this.leftRule = leftRule;
        this.infix = infix;
        this.rightRule = rightRule;
        this.factory = factory;
    }

    @Override
    public NodeResult lex(String input) {
        final var index = input.indexOf(this.infix);
        if (index == -1)
            return this.factory.fromStringErr("Infix '" + this.infix + "' not present", input);

        final var left = input.substring(0, index);
        final var right = input.substring(index + this.infix.length());
        return this.getNodeResult(this.leftRule.lex(left), right);
    }

    private NodeResult getNodeResult(NodeResult leftResult, String slice) {
        return leftResult.mergeResult(() -> this.rightRule.lex(slice));
    }

    @Override
    public StringResult generate(Node node) {
        return this.appendInfix(node, this.leftRule.generate(node));
    }

    private StringResult appendInfix(Node node, StringResult result) {
        return this.appendRight(node, result.appendSlice(this.infix));
    }

    private StringResult appendRight(Node node, StringResult result1) {
        return result1.appendResult(() -> this.rightRule.generate(node));
    }
}
